kernel/mman.c

```diff
@@ -426,6 +429,90 @@ static uintptr_t _mman_find_gap(
     return addr;
 }
 
+#define _MMAN_MPROTECT_PAGES(MMAN, ADDR, LEN, PROT)            \
+    {                                                          \
+        if (myst_tcall_mprotect(ADDR, LEN, PROT))              \
+        {                                                      \
+            _mman_set_err(MMAN, "mprotect tcall failed");      \
+            ret = -EINVAL;                                     \
+            goto done;                                         \
+        }                                                      \
+        memset(                                                \
+            (MMAN)->prot_vector +                              \
+                ((uintptr_t)ADDR - (MMAN)->start) / PAGE_SIZE, \
+            PROT,                                              \
+            (LEN) / PAGE_SIZE);                                \
+    }
+
+/* set each page within the range's permission tracking as prot*/
+#define _MMAN_SET_PAGES_PROT(MMAN, ADDR, LEN, PROT)            \
+    {                                                          \
+        memset(                                                \
+            (MMAN)->prot_vector +                              \
+                ((uintptr_t)ADDR - (MMAN)->start) / PAGE_SIZE, \
+            PROT,                                              \
+            (LEN) / PAGE_SIZE);                                \
+    }
+
+/* check and get permission tracking, if insisitent, return -1*/
+static int _mman_get_prot(
```

Comment:
Very nice optimization. One optimization might be to use __uint128_t iteration instead, which is even faster. Also I wonder if we could generalize this into a string function, like memcmp_char()?

Anyway, this would all be the future.